src/field/fieldgenerators.cxx

```diff
@@ -268,13 +269,148 @@ std::shared_ptr<FieldGenerator> FieldTanhHat::clone(const list<std::shared_ptr<F
 }
 
-BoutReal FieldTanhHat::generate(double x, double y, double z, double t) {
+BoutReal FieldTanhHat::generate(double x, double y, double z, double t, const DataIterator &i, Mesh *localmesh) {
   // The following are constants
-  BoutReal w = width    ->generate(0,0,0,0);
-  BoutReal c = center   ->generate(0,0,0,0);
-  BoutReal s = steepness->generate(0,0,0,0);
+  BoutReal w = width    ->generate(0,0,0,0,i,localmesh);
+  BoutReal c = center   ->generate(0,0,0,0,i,localmesh);
+  BoutReal s = steepness->generate(0,0,0,0,i,localmesh);
   return 0.5*(
-                 tanh( s*(X->generate(x,y,z,t) - (c - 0.5*w)) )
-               - tanh( s*(X->generate(x,y,z,t) - (c + 0.5*w)) )
+                 tanh( s*(X->generate(x,y,z,t,i,localmesh) - (c - 0.5*w)) )
+               - tanh( s*(X->generate(x,y,z,t,i,localmesh) - (c + 0.5*w)) )
              );
 }
+
+BoutReal FieldRealY::generate(double x, double y, double z, double t,
+                              const DataIterator &i, Mesh *localmesh) {
+  BoutReal *offset = doCache(localmesh);
+  int gny = localmesh->GlobalNy - mesh->ystart * 2;
+  int iy = y / 2 / PI * gny * 2 + 10.5;
+  BoutReal res;
+  Field2D &dy = localmesh->coordinates()->dy;
+  res = offset[i.x];
+  for (int y = 0; y < i.y; ++y) {
+    res += dy(i.x, y);
+  }
+  if (iy % 2) { // centered
+    res -= dy[i] * .5;
+  } else {
+    // res+=dy[i];
+  }
+  // printf("%g %8g %4d %4d %6g\n",offset[i.x],y/2/PI*gny*2,i.y,iy,res);
+  return res;
+}
+
+BoutReal *FieldRealY::doCache(Mesh *localmesh) {
+  auto it = cache.begin();
+  for (const auto &cm : cached) {
+    if (cm == localmesh) {
+      return *it;
+    }
+    ++it;
+  }
+  cached.push_back(localmesh);
+  BoutReal *offset = (BoutReal *)malloc(sizeof(BoutReal) * localmesh->LocalNx);
+  BoutReal tmp[localmesh->LocalNx];
+  BoutReal dummy1, dummy2;
+  if (localmesh->firstY()) {
+    for (int x = 0; x < localmesh->LocalNx; ++x) {
+      offset[x] = 0;
+      for (int y = localmesh->ystart - 1; y >= 0; --y) {
+        offset[x] -= localmesh->coordinates()->dy(x, y);
+      }
+    }
+  } else {
+    localmesh->wait(localmesh->receiveFromProc(localmesh->getXProcIndex(),
+                                               localmesh->getYProcIndex() - 1, offset,
+                                               localmesh->LocalNx, 666));
+    // printf("%d: receive %g\n",localmesh->getYProcIndex(),offset[localmesh->xstart]);
+  }
+  if (!localmesh->lastY()) {
+    for (int x = 0; x < localmesh->LocalNx; ++x) {
+      tmp[x] = offset[x];
+      for (int y = 0; y < localmesh->LocalNy - localmesh->ystart * 2; ++y) {
+        tmp[x] += localmesh->coordinates()->dy(x, y);
+      }
+    }
+    // printf("%d: send %g\n",localmesh->getYProcIndex(),tmp[localmesh->xstart]);
+    MPI_Request out =
+        localmesh->sendToProc(localmesh->getXProcIndex(), localmesh->getYProcIndex() + 1,
+                              tmp, localmesh->LocalNx, 666);
+    MPI_Status stat;
+    MPI_Wait(&out, &stat);
```

Comment:
Something seems to be off with the MPI calls - during deallocation I get segfaults in
```
# 3  BoutMesh::clear_handles (this=this@entry=0x555555bfa1d0) at boutmesh.cxx:1918
1918        comm_list.pop_front();
```

Is there anything wrong with they way I have used sendToProc/receiveFromProc?
I was further surprised that both calls where non-blocking - for non blocking calls I would have expected Isend and Ireceive. That was also not mentioned in the documentation, as far as I could see ...

Reply:
Ignore above comment - segfault seems to be unrelated ....